### PR TITLE
perf(nta) NTA scheduling performance improvement by optimizing hyperNodeGradientFn

### PR DESCRIPTION
**Pull request by 3th4novo**
#### What type of PR is this?

area/performance kind/feature

#### What this PR does / why we need it:

Currently, the network topology-aware plugin iterates from bottom to top in a hypernode tree and caused inefficiency. For detailed description please review issue [#4899](https://github.com/volcano-sh/volcano/issues/4899#event-21805727490) . This PR introduces HyperNode pre-filtering logic within hyperNodeGradientFn. Specifically, we only keep a hypernode to eligibleHyperNodes when minResource is less than or equal to the idle resource of a hypernode, or the future idle resource of a hypernode.

Note that here minResource can refer to both the minimum required resource of a job or a sub job.

The PR also modifies sub_job_info.go for sub job minReource calculation purposes. Right now a sub job is constrained to gang scheduling, so the minimum required resource of a sub job can be considered as the total initResReq of each pending task in the sub job.

```go
func (sji *SubJobInfo) GetMinResources() *Resource {
	totalResource := EmptyResource()
	for _, task := range sji.TaskStatusIndex[Pending] {
		totalResource.Add(task.InitResreq)
	}
	return totalResource
}
``` 

Improvement: Under the setup as in issue 4899, when we have a subgroup for each pod, resulting in 1000 subgroups, the total scheduling duration can last about 1 minute. This number is optimized to a few seconds in this PR. When we build 100 subgroups, each containing 10 pods, scheduling duration is optimized by about 0.5 seconds, from about 6.3 seconds to 5.9 seconds. By observation, I believe the differences in the number of subgroups mainly contribute to scheduling duration before optimization. The main reason is that the scheduler will have to perform more dry run operations.

#### Which issue(s) this PR fixes:

#4899 issue 1

#### Special notes for your reviewer:

This PR is based on [#4889](https://github.com/volcano-sh/volcano/pull/4889) so it should be merged after #4889 is merged.

Feel free to modify TestHyperNodeGradientPreFiltering for correctness checks and BenchmarkHyperNodeGradientFnPerformance for performance benchmarks. 

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: pkg/scheduler/plugins/network-topology-aware/network_topology_aware.go [hunk 1]
```diff
@@ -380,7 +540,15 @@ func (nta *networkTopologyAwarePlugin) hyperNodeGradientFn(ssn *framework.Sessio
...
+			if minResource != nil {
+				hnResourceStatus := nta.hyperNodeResourceCache[current.Name]
+				if hnResourceStatus != nil {
+					if minResource.LessEqual(hnResourceStatus.idle, api.Zero) ||
+						minResource.LessEqual(hnResourceStatus.futureIdle, api.Zero) {
+						eligibleHyperNodes[current.Tier()] = append(eligibleHyperNodes[current.Tier()], current)
+					}
+				}
+			}
 		}
```
[the remark this answers is not in the document]
Comment (3th4novo) [reply pkg/scheduler/plugins/network-topology-aware/network_topology_aware.go]: I have corrected it.

File: pkg/scheduler/plugins/network-topology-aware/network_topology_aware_test.go [hunk 2]
```diff
@@ -1928,10 +2109,1424 @@ func parseTask(jobInfoMap map[api.JobID]*api.JobInfo) *api.TaskInfo {
...
+		if subJob.UID == "ns1/pg1/task1-worker" {
+			return subJob
+		}
+	}
+	return nil
+}
+
+// TestHyperNodeGradientPreFiltering tests the pre-filtering logic in hyperNodeGradientFn.
+// It verifies that HyperNodes are correctly filtered based on idle and futureIdle resources.
+func TestHyperNodeGradientPreFiltering(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (3th4novo) [reply pkg/scheduler/plugins/network-topology-aware/network_topology_aware_test.go:3389]: I have added two more test cases where minResource == nil, one for job and another one for sub job.

File: pkg/scheduler/plugins/network-topology-aware/network_topology_aware.go [hunk 3]
```diff
@@ -229,10 +298,31 @@ func (nta *networkTopologyAwarePlugin) OnSessionOpen(ssn *framework.Session) {
...
+		DeallocateFunc: func(event *framework.Event) {
+			task := event.Task
+			node := task.NodeName
+			for hyperNode := range ssn.HyperNodes {
+				if ssn.RealNodesSet[hyperNode].Has(node) {
+					nta.hyperNodeResourceCache[hyperNode].used.Sub(task.Resreq)
+				}
+			}
+		},
+	})
```
[the remark this answers is not in the document]
Comment (3th4novo) [reply pkg/scheduler/plugins/network-topology-aware/network_topology_aware.go:338]: will fix in the following review.

[the remark this answers is not in the document]
Comment (3th4novo) [reply pkg/scheduler/plugins/network-topology-aware/network_topology_aware.go:338]: code updated, will review again.

File: pkg/scheduler/actions/allocate/allocate_test.go [hunk 4]
```diff
@@ -5182,3 +5184,169 @@ func TestAllocateWithPartitionPolicyNetworkTopology(t *testing.T) {
...
+
+	for _, job := range ssn.Jobs {
+		topHyperNode := ssn.HyperNodes[framework.ClusterTopHyperNode]
+		gradient := ssn.HyperNodeGradientForJobFn(job, topHyperNode)
+		klog.Infof("\nGradient for Job %s: %d tiers", job.UID, len(gradient))
+		for tierIdx, tierHNs := range gradient {
+			klog.Infof("  Tier %d: %d hyperNodes", tierIdx+1, len(tierHNs))
+		}
+	}
+}
```
[the remark this answers is not in the document]
Comment (3th4novo) [reply pkg/scheduler/actions/allocate/allocate_test.go:5715]: fixed

File: pkg/scheduler/plugins/network-topology-aware/network_topology_aware.go [hunk 5]
```diff
@@ -562,7 +571,18 @@ func (nta *networkTopologyAwarePlugin) hyperNodeGradientFn(ssn *framework.Sessio
...
+			} else {
+				hnResourceStatus, found := nta.hyperNodeResourceCache[current.Name]
+				if !found {
+					klog.Warningf("Resource status for hypernode %s not found in cache, skipping pre-filtering for it.", current.Name)
+					eligibleHyperNodes[current.Tier()] = append(eligibleHyperNodes[current.Tier()], current)
+				} else if minResource.LessEqual(hnResourceStatus.idle, api.Zero) ||
+					minResource.LessEqual(hnResourceStatus.futureIdle, api.Zero) {
+					eligibleHyperNodes[current.Tier()] = append(eligibleHyperNodes[current.Tier()], current)
+				}
+			}
```
[the remark this answers is not in the document]
Comment (3th4novo) [reply pkg/scheduler/plugins/network-topology-aware/network_topology_aware.go]: fixed

File: pkg/scheduler/plugins/network-topology-aware/network_topology_aware.go [hunk 6]
```diff
@@ -562,7 +571,14 @@ func (nta *networkTopologyAwarePlugin) hyperNodeGradientFn(ssn *framework.Sessio
 		processQueue = processQueue[1:]
 
 		if current.Tier() <= highestAllowedTier {
-			eligibleHyperNodes[current.Tier()] = append(eligibleHyperNodes[current.Tier()], current)
+			hnResourceStatus, found := nta.hyperNodeResourceCache[current.Name]
+			if !found {
+				klog.Warningf("Resource status for hypernode %s not found in cache, skipping pre-filtering for it.", current.Name)
+				eligibleHyperNodes[current.Tier()] = append(eligibleHyperNodes[current.Tier()], current)
+			} else if minResource.LessEqual(hnResourceStatus.idle, api.Zero) ||
+				minResource.LessEqual(hnResourceStatus.futureIdle, api.Zero) {
```
[the remark this answers is not in the document]
Comment (3th4novo) [reply pkg/scheduler/plugins/network-topology-aware/network_topology_aware.go]: fixed

File: pkg/scheduler/plugins/network-topology-aware/network_topology_aware.go [hunk 7]
```diff
@@ -543,7 +551,10 @@ func (nta *networkTopologyAwarePlugin) batchNodeOrderFnForNetworkAwarePods(ssn *
 	return nodeScores, nil
 }
 
-func (nta *networkTopologyAwarePlugin) hyperNodeGradientFn(ssn *framework.Session, hyperNode *api.HyperNodeInfo, highestAllowedTier int, allocatedHyperNode string) ([][]*api.HyperNodeInfo, error) {
+// minResource is the minimum required resource for the job/subjob to be scheduled.
```
Comment (JesseStutler) [inline pkg/scheduler/plugins/network-topology-aware/network_topology_aware.go]: The explanation should cover the entire function and the purpose of all its parameters, not just the explanation of a single parameter.

File: pkg/scheduler/actions/allocate/allocate_test.go [hunk 8]
```diff
@@ -5182,3 +5184,169 @@ func TestAllocateWithPartitionPolicyNetworkTopology(t *testing.T) {
 		})
 	}
 }
+
+// TestHyperNodeGradientFnPerformance tests the performance optimization
+// of hyperNodeGradientFn with SubGroup policy in large-scale cluster scenarios.
+func TestHyperNodeGradientFnPerformance(t *testing.T) {
```
Comment (JesseStutler) [inline pkg/scheduler/actions/allocate/allocate_test.go]: I think the performance test should use `*testing.B` instead, I didn't find any validation for using `*testing.T`

Comment (3th4novo) [reply pkg/scheduler/actions/allocate/allocate_test.go]: Fixed, ty.

File: pkg/scheduler/plugins/network-topology-aware/network_topology_aware.go (hunk 7 above)
Comment (3th4novo) [reply pkg/scheduler/plugins/network-topology-aware/network_topology_aware.go]: Fixed, ty.

File: pkg/scheduler/actions/allocate/allocate_test.go (hunk 8 above)
Comment (JesseStutler) [reply pkg/scheduler/actions/allocate/allocate_test.go]: @3th4novo I mean you should use testing.B to run b.N times, and the testing struct may should follow the style like `BenchmarkAllocate`

Comment (3th4novo) [reply pkg/scheduler/actions/allocate/allocate_test.go]: Please refer to the latest commit. Let me know if there are any changes needed.

File: pkg/scheduler/api/sub_job_info.go [hunk 9]
```diff
@@ -261,3 +261,14 @@ func (sji *SubJobInfo) AllocatedTaskNum() int32 {
 func (sji *SubJobInfo) CloneStatusFrom(source *SubJobInfo) {
 	sji.AllocatedHyperNode = source.AllocatedHyperNode
 }
+
+// GetMinResources The current sub job is constrained to gang scheduling,
+// so the minimum required resource of a sub job can be considered as
+// the total initResReq of each task in the sub job.
+func (sji *SubJobInfo) GetMinResources() *Resource {
```
Comment (Tau721) [inline pkg/scheduler/api/sub_job_info.go:269]: This function sums up the `InitResreq` of all tasks of a subjob and uses it as the minimal resource requirement to filter hypernodes. However, what if some tasks of this subjob have already been running? I think the resource requirements of these tasks should be ignored.

Comment (3th4novo) [reply pkg/scheduler/api/sub_job_info.go:269]: I do think that here we should still consider the requirements of the pending tasks. Actually I used the InitResreq of all pending tasks of a subjob here in the previous versions, so that the function would do the correct calculation under this circumstance. Do you suggest I implement this way?

Comment (3th4novo) [reply pkg/scheduler/api/sub_job_info.go:269]: @ouyangshengjia @JesseStutler What do you guys think of the desired implementation here, since it's a new design?

Comment (JesseStutler) [reply pkg/scheduler/api/sub_job_info.go:269]: I think @Tau721 is correct, we should ignore the running pods, however, the semantics of this function are rather strange, minResources is simply meant to schedule minAvailable pods.

Comment (JesseStutler) [reply pkg/scheduler/api/sub_job_info.go:269]: I remember that we also contains some fields like min number count in subjob, so we should only consider the subjob level's min number count

Comment (3th4novo) [reply pkg/scheduler/api/sub_job_info.go:269]: I agree with you on only counting the initresreq of the pending tasks of a sub job and I have updated it in the latest commit. However, currently the constraint of a sub job is that MinAvailable == size of subjob, therefore I believe my implementation makes sense.

File: pkg/scheduler/plugins/network-topology-aware/network_topology_aware.go [hunk 10]
```diff
@@ -271,7 +277,8 @@ func (nta *networkTopologyAwarePlugin) OnSessionOpen(ssn *framework.Session) {
 
 	ssn.AddHyperNodeGradientForJobFn(nta.Name(), func(job *api.JobInfo, hyperNode *api.HyperNodeInfo) [][]*api.HyperNodeInfo {
 		if hardMode, highestAllowedTier := job.IsHardTopologyMode(); hardMode {
-			result, err := nta.hyperNodeGradientFn(ssn, hyperNode, highestAllowedTier, job.AllocatedHyperNode)
+			jobMinResource := job.GetMinResources()
```
Comment (JesseStutler) [inline pkg/scheduler/plugins/network-topology-aware/network_topology_aware.go:280]: So you should also consider rescheduling scenarios for job, and I feel the job scenario is even more complex, for example, evicted pods might belong to different subjobs, and these subjobs are distributed under different hypernodes. Therefore, we cannot use simple summation to calculate minResources.

File: pkg/scheduler/api/sub_job_info.go (hunk 9 above)
Comment (JesseStutler) [reply pkg/scheduler/api/sub_job_info.go:269]: @3th4novo Also please add test cases to go through the whole allocate action for rescheduling scenario, you can add cases in allocate_test.go, e.g., add in `TestAllocateWithPartitionPolicyNetworkTopology`

File: pkg/scheduler/plugins/network-topology-aware/network_topology_aware.go (hunk 10 above)
Comment (ouyangshengjia) [reply pkg/scheduler/plugins/network-topology-aware/network_topology_aware.go:280]: Indeed, in scenarios where podgroups are partially running or in failure rescheduling scenarios (such as `RestartTask`/`RestartPartition` of vcjob), the pruning conditions here become very complex. I was wondering whether this scheduling pruning optimization could be applied only to podgroups that are **completely pending**, while allowing partially running podgroups to proceed without restriction.

Comment (JesseStutler) [reply pkg/scheduler/plugins/network-topology-aware/network_topology_aware.go:280]: Great, this is also a method, but I'm wondering if a hypernode that can't accommodate the total resources calculated through simple addition is necessarily not eligible as a candidate. Since we're doing pre-predicate here, even if topology isn't considered here, it will still go through the predicate later, but at least it can exclude hypernodes that are definitely resource-insufficient

Comment (ouyangshengjia) [reply pkg/scheduler/plugins/network-topology-aware/network_topology_aware.go:280]: For job-level pruning optimization, **do not consider the failure rescheduling scenarios** at the beginning:
When a job meets the minavailable condition, only one pod needs to be scheduled subsequently to complete this round of scheduling. Therefore, pruning evaluation is only required when the job is completely pending.

**If the failure rescheduling scenarios is considered**, the job may be in a state where some pods are running but the minavailable condition **is not yet met**. In this case, if pruning evaluation is performed, the resources required for the minavailable condition need to be calculated by deducting the resources occupied by the currently running pods, making the calculation condition more complex. Considering that the rescheduling scenario may not be common, **it is recommended that pruning be not performed in this scenario.**

Comment (3th4novo) [reply pkg/scheduler/plugins/network-topology-aware/network_topology_aware.go:280]:
```go
if allocatedHyperNode != "" {
    eligibleHyperNodes[current.Tier()] = append(eligibleHyperNodes[current.Tier()], current)
}
```

Please check line 584-585. Under rescheduling scenarios, the job/subjob would be allocated to some hypernode and it will skip pre-filtering.

Comment (ouyangshengjia) [reply pkg/scheduler/plugins/network-topology-aware/network_topology_aware.go:280]: @3th4novo You can add comments in code to explain why your pre-filtering is only done on this branch.

Comment (3th4novo) [reply pkg/scheduler/plugins/network-topology-aware/network_topology_aware.go:280]: done.

File: pkg/scheduler/api/sub_job_info.go (hunk 9 above)
Comment (3th4novo) [reply pkg/scheduler/api/sub_job_info.go:269]: done.

File: pkg/scheduler/plugins/network-topology-aware/network_topology_aware.go [hunk 11]
```diff
@@ -562,7 +580,20 @@ func (nta *networkTopologyAwarePlugin) hyperNodeGradientFn(ssn *framework.Sessio
 		processQueue = processQueue[1:]
 
 		if current.Tier() <= highestAllowedTier {
```
Comment (ouyangshengjia) [inline pkg/scheduler/plugins/network-topology-aware/network_topology_aware.go]: It is recommended to extract a filter function as follows:
```golang
if nta.isEligibleHyperNode(current, ...) {
    eligibleHyperNodes[current.Tier()] = append(eligibleHyperNodes[current.Tier()], current)
}

func (nta *networkTopologyAwarePlugin)isEligibleHyperNode(hn *api.HyperNodeInfo, ...) bool {
    if hn.Tier() > highestAllowedTier {
        return false // the tier should not exceed the highest allowed
    }

    if allocatedHyperNode != "" {
        return true // skip pre-filtering in partially running scenarios
    }
    stats, found := nta.hyperNodeResourceCache[current.Name]
    if !found {
        return true // skip pre-filtering due to insufficient conditions
    }
    if minResource.LessEqual(stats.idle, api.Zero) || minResource.LessEqual(stats.futureIdle, api.Zero) {
        return true
    }
    return false
}
```

File: pkg/scheduler/plugins/network-topology-aware/network_topology_aware.go [hunk 12]
```diff
@@ -543,7 +551,17 @@ func (nta *networkTopologyAwarePlugin) batchNodeOrderFnForNetworkAwarePods(ssn *
...
-func (nta *networkTopologyAwarePlugin) hyperNodeGradientFn(ssn *framework.Session, hyperNode *api.HyperNodeInfo, highestAllowedTier int, allocatedHyperNode string) ([][]*api.HyperNodeInfo, error) {
+// hyperNodeGradientFn computes network topology gradients by performing BFS traversal from the given HyperNode,
+// filtering and grouping HyperNodes by tier based on resource availability and topology constraints.
+// It returns HyperNodes organized in ascending tier order (lower tiers represent closer network proximity).
+//
+// Parameters:
+//   - ssn: scheduling session containing all HyperNode information and cluster state
+//   - hyperNode: starting HyperNode for the search, typically the root of available HyperNode subtree
+//   - highestAllowedTier: maximum allowed topology tier to limit search scope
+//   - allocatedHyperNode: previously allocated HyperNode name for rescheduling scenarios (empty for initial scheduling)
```
Comment (ouyangshengjia) [inline pkg/scheduler/plugins/network-topology-aware/network_topology_aware.go]: Just a suggestion, I think "partially running" is more accurate than "rescheduling"

Comment (3th4novo) [reply pkg/scheduler/plugins/network-topology-aware/network_topology_aware.go]: fixed

File: pkg/scheduler/plugins/network-topology-aware/network_topology_aware.go (hunk 11 above)
Comment (3th4novo) [reply pkg/scheduler/plugins/network-topology-aware/network_topology_aware.go]: fixed

File: pkg/scheduler/plugins/network-topology-aware/network_topology_aware.go [hunk 13]
```diff
@@ -590,6 +608,25 @@ func (nta *networkTopologyAwarePlugin) hyperNodeGradientFn(ssn *framework.Sessio
...
+
+	if allocatedHyperNode != "" {
+		return true // skip pre-filtering in partially running scenarios
+	}
+
+	hnResourceStatus, found := nta.hyperNodeResourceCache[hn.Name]
+	if !found {
+		return true // Resource status for hypernode not found in cache, skipping pre-filtering for it.
+	}
+	if minResource == nil || minResource.LessEqual(hnResourceStatus.idle, api.Zero) || minResource.LessEqual(hnResourceStatus.futureIdle, api.Zero) {
```
Comment (ouyangshengjia) [inline pkg/scheduler/plugins/network-topology-aware/network_topology_aware.go]: In which scenario will `minResource` be `nil`?

Comment (3th4novo) [reply pkg/scheduler/plugins/network-topology-aware/network_topology_aware.go]: I did that according to an older gemini suggestion... mainly for safety reasons. But in fact both the GetMinResources method from a job and sub job wouldn't introduce a nil value. Let me get rid of this value check.
